File: leetcode/ransom-note.go
package main

import (
	"fmt"
)

func canConstruct(ransomNote string, magazine string) bool {
	count := make(map[string]int)

	for _, s := range magazine {
		count[string(s)] += 1
	}

	for _, s := range ransomNote {
		if val, ok := count[string(s)]; ok {
			if val <= 0 {
				return false
			}
			count[string(s)] -= 1
		} else {
			return false
		}
	}
	return true

}

func main() {
	// ransomNote := "aa"
	// magazine := "aab"
	ransomNote := "aa"
	magazine := "ab"

	fmt.Println(canConstruct(ransomNote, magazine))
}

File: leetcode/ransom-note.py


class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        reserved = []
        magazine = list(magazine)
        for s in ransomNote:
            if s not in magazine:
                return False
            magazine.remove(s)
        return True


if __name__ == "__main__":
    solution = Solution()
    ransomNote = "a"
    magazine = "b"
    print(solution.canConstruct(ransomNote, magazine))

    ransomNote = "aa"
    magazine = "aab"

    print(solution.canConstruct(ransomNote, magazine))

    ransomNote = "aa"
    magazine = "ab"
    print(solution.canConstruct(ransomNote, magazine))

